File: temp/BTree_.go
package main

import "fmt"

type Node struct {
	leaf     bool
	keys     []int
	children []*Node
}

type BTree struct {
	root *Node
	t    int // for each non-root node, t-1 <= len(keys) <= 2*t-1
}

func NewTree(t int) *BTree {
	return &BTree{root: &Node{leaf: true}, t: t}
}

func (bt *BTree) Insert(key int) {
	root, t := bt.root, bt.t
	if len(root.keys) == (2*t - 1) {
		rootN := &Node{}
		bt.root = rootN
		rootN.children = append(rootN.children, root)
		bt.splitChild(rootN, 0)
		bt.insertNonFull(rootN, key)
	} else {
		bt.insertNonFull(root, key)
	}
}

func (bt *BTree) splitChild(x *Node, idx int) {
	from := x.children[idx]
	dest := &Node{leaf: from.leaf}

	// insert new node:from to n.children at pos idx+1, and shift eles from idx+1 to right by 1
	tmp := make([]*Node, 0)
	tmp = append(tmp, x.children[:idx+1]...)
	tmp = append(tmp, dest)
	tmp = append(tmp, x.children[idx+1:]...)
	x.children = tmp
	//tmp := append([]*Node{}, x.children[idx+1:]...)
	//x.children = x.children[:idx+1]
	//x.children = append(x.children, dest)
	//x.children = append(x.children, tmp...)

	// promote the key at pos:t-1 from node:from to x.keys at pos:idx
	tmp2 := make([]int, 0)
	tmp2 = append(tmp2, x.keys[:idx]...)
	tmp2 = append(tmp2, from.keys[bt.t-1])
	tmp2 = append(tmp2, x.keys[idx:]...)
	x.keys = tmp2
	//tmp2 := append([]int{}, x.keys[idx:]...)
	//x.keys = x.keys[:idx]
	//x.keys = append(x.keys, from.keys[bt.t-1])
	//x.keys = append(x.keys, tmp2...)

	// move keys from idx:t in node:from to node:dest
	dest.keys = append(dest.keys, from.keys[bt.t:]...)
	// remove keys from idx:t-1 in node:from
	from.keys = from.keys[:bt.t-1]

	if !from.leaf {
		// move half children from node:from to node:dest
		dest.children = append(dest.children, from.children[bt.t:]...)
		// clear the half children that were just moved in node:from
		from.children = from.children[:bt.t]
	}
}

func (bt *BTree) insertNonFull(x *Node, key int) {
	idx := len(x.keys) - 1
	if x.leaf {
		for idx >= 0 && key < x.keys[idx] {
			idx--
		}
		tmp := make([]int, 0)
		tmp = append(tmp, x.keys[:idx+1]...)
		tmp = append(tmp, key)
		tmp = append(tmp, x.keys[idx+1:]...)
		x.keys = tmp
		//tmp := append([]int{}, x.keys[idx+1:]...)
		//x.keys = x.keys[:idx+1]
		//x.keys = append(x.keys, key)
		//x.keys = append(x.keys, tmp...)
	} else {
		for idx >= 0 && key < x.keys[idx] {
			idx--
		}
		idx++
		if len(x.children[idx].keys) == (2*bt.t - 1) {
			bt.splitChild(x, idx)
			if key > x.keys[idx] {
				idx++
			}
		}
		bt.insertNonFull(x.children[idx], key)
	}
}

// Lookup searches for a key in the B-tree. Returns true if the key is found, along with the associated value.
func (bt *BTree) Lookup(key int) (bool, int) {
	return bt.lookupKey(bt.root, key)
}

// Helper function for key lookup
func (bt *BTree) lookupKey(node *Node, key int) (bool, int) {
	i := 0
	for i < len(node.keys) && key > node.keys[i] {
		i++
	}

	if i < len(node.keys) && key == node.keys[i] {
		// Key found in the current node
		return true, key
	} else if node.leaf {
		// Key not found in a leaf node
		return false, 0
	} else {
		// Recursively search in the appropriate children
		return bt.lookupKey(node.children[i], key)
	}
}

func (bt *BTree) PrintTree(n *Node) {
	q := []*Node{n}
	for len(q) > 0 {
		size := len(q)
		for i := 0; i < size; i++ {
			x := q[0]
			q = q[1:]
			fmt.Print(x.keys, "------")
			for _, child := range x.children {
				q = append(q, child)
			}
		}
		fmt.Println()
	}
}

func main() {
	btree := NewTree(2) // Create a B-tree with a minimum degree of 2

	keys := []int{29, 41, 44, 62, 46, 49, 27, 76, 91, 30, 100, 47, 34, 53, 9, 45}
	//keys := []int{29, 41, 44, 62, 46}
	for _, key := range keys {
		btree.Insert(key)
	}

	btree.PrintTree(btree.root)
}

File: Archive/BTree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class BTreeNode {
    boolean leaf;
    List<Integer> keys;
    List<BTreeNode> children;

    public BTreeNode(boolean leaf) {
        this.leaf = leaf;
        this.keys = new ArrayList<>();
        this.children = new ArrayList<>();
    }
}

public class BTree {
    private BTreeNode root;
    private int t; // Minimum degree

    public BTree(int t) {
        this.root = new BTreeNode(true);
        this.t = t;
    }

    public void insert(int key) {
        BTreeNode r = root;
        if (r.keys.size() == (2 * t - 1)) {
            BTreeNode s = new BTreeNode(false);
            root = s;
            s.children.add(r);
            splitChild(s, 0);
            insertNonFull(s, key);
        } else {
            insertNonFull(r, key);
        }
    }

    private void insertNonFull(BTreeNode x, int key) {
        int i = x.keys.size() - 1;
        if (x.leaf) {
            while (i >= 0 && key < x.keys.get(i)) {
                i--;
            }
            x.keys.add(i + 1, key);
        } else {
            while (i >= 0 && key < x.keys.get(i)) {
                i--;
            }
            i++;
            if (x.children.get(i).keys.size() == (2 * t - 1)) {
                splitChild(x, i);
                if (key > x.keys.get(i)) {
                    i++;
                }
            }
            insertNonFull(x.children.get(i), key);
        }
    }

    private void splitChild(BTreeNode x, int i) {
        BTreeNode y = x.children.get(i);
        BTreeNode z = new BTreeNode(y.leaf);

        x.children.add(i + 1, z);
        x.keys.add(i, y.keys.get(t - 1));

        z.keys.addAll(y.keys.subList(t, 2 * t - 1));
        y.keys.subList(t - 1, 2 * t - 1).clear();

        if (!y.leaf) {
            z.children.addAll(y.children.subList(t, 2 * t));
            y.children.subList(t, 2 * t).clear();
        }
    }

    public void print() {
        print(root, 0);
    }

    private void print(BTreeNode node, int level) {
        System.out.println("Level " + level + ": " + node.keys);
        if (!node.leaf) {
            for (BTreeNode child : node.children) {
                print(child, level + 1);
            }
        }
    }
    public void printTree(BTreeNode node) {
        if (node == null) {
            return;
        }

        Queue<BTreeNode> queue = new LinkedList<>();
        queue.offer(node);

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                BTreeNode x = queue.poll();
                if (x != null) {
                    // Print keys
                    for (int key : x.keys) {
                        System.out.print(key + " ");
                    }
                    System.out.print("---");

                    // Enqueue children
                    if (!x.leaf) {
                        for (BTreeNode child : x.children) {
                            queue.offer(child);
                        }
                    }
                }
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        BTree bTree = new BTree(3);

        int[] keys = {34, 11, 76, 53, 29, 48, 65, 95, 81, 92, 68, 59, 87, 20, 45, 26, 83, 70, 37, 7, 17, 73, 42, 96, 23, 58, 8, 50, 94, 61};
//        int[] keys = {29, 41, 44, 62, 46};

        for (int key : keys) {
            bTree.insert(key);
        }

        System.out.println("B-tree structure:");
        bTree.printTree(bTree.root);
    }
}
